src/flamenco/runtime/program/fd_bpf_loader_serialization.c

```diff
@@ -4,92 +4,231 @@
       }
 
-      ulong aligned_acc_data_len = fd_ulong_align_up(acc_data_len, 8);
-
-      serialized_size += sizeof(uchar)  // is_signer
-          + sizeof(uchar)               // is_writable
-          + sizeof(uchar)               // is_executable
-          + sizeof(uint)                // original_data_len
-          + sizeof(fd_pubkey_t)         // key
-          + sizeof(fd_pubkey_t)         // owner
-          + sizeof(ulong)               // lamports
-          + sizeof(ulong)               // data_len
-          + aligned_acc_data_len
-          + MAX_PERMITTED_DATA_INCREASE
-          + sizeof(ulong);              // rent_epoch
+      serialized_size += sizeof(uchar)               // is_signer
+                       + sizeof(uchar)               // is_writable
+                       + sizeof(uchar)               // executable
+                       + sizeof(uint)                // original_data_len
+                       + sizeof(fd_pubkey_t)         // key
+                       + sizeof(fd_pubkey_t)         // owner
+                       + sizeof(ulong)               // lamports
+                       + sizeof(ulong)               // data len
+                       + MAX_PERMITTED_DATA_INCREASE
+                       + sizeof(ulong);              // rent_epoch
+      if( copy_account_data ) {
+        serialized_size += fd_ulong_align_up( acc_data_len, FD_BPF_ALIGN_OF_U128 );
+      } else {
+        serialized_size += FD_BPF_ALIGN_OF_U128;
+      }
     }
   }
 
-  serialized_size += sizeof(ulong)
-      + ctx.instr->data_sz
-      + sizeof(fd_pubkey_t);
-  uchar * serialized_params = fd_valloc_malloc( ctx.valloc, 8UL, serialized_size );
-  uchar * serialized_params_start = serialized_params;
+  serialized_size += sizeof(ulong)        // data len 
+                  +  ctx.instr->data_sz
+                  +  sizeof(fd_pubkey_t); // program id
+
+  uchar * serialized_params            = fd_valloc_malloc( ctx.valloc, FD_BPF_ALIGN_OF_U128, serialized_size );
+  uchar * serialized_params_start      = serialized_params;
+  uchar * curr_serialized_params_start = serialized_params;
 
   FD_STORE( ulong, serialized_params, ctx.instr->acct_cnt );
   serialized_params += sizeof(ulong);
 
-  for( ushort i = 0; i < ctx.instr->acct_cnt; i++ ) {
-    uchar acc_idx = instr_acc_idxs[i];
-    fd_pubkey_t * acc = &txn_accs[acc_idx];
+  /* Second pass over the account is to serialize into the buffer. */
+  for( ushort i=0; i<ctx.instr->acct_cnt; i++ ) {
+    uchar         acc_idx = instr_acc_idxs[i];
+    fd_pubkey_t * acc     = &txn_accs[acc_idx];
 
     if( FD_UNLIKELY( acc_idx_seen[acc_idx] && dup_acc_idx[acc_idx] != i ) ) {
-      // Duplicate
-      FD_STORE( ulong, serialized_params, 0 );
+      /* Duplicate. Store 8 byte buffer to maintain alignment but store the
+         account index in the first byte.*/
+      FD_STORE( ulong, serialized_params, 0UL );
       FD_STORE( uchar, serialized_params, (uchar)dup_acc_idx[acc_idx] );
```

Comment:
Are these two the correct way round? Agave serializes them in little-endian: https://github.com/anza-xyz/agave/blob/36323b6dcd3e29e4d6fe6d73d716a3f33927148b/programs/bpf_loader/src/serialization.rs#L488-L489